Search by app profile for managing releases

mkangia opened the pull request:
[Ticket](https://dimagi-dev.atlassian.net/browse/ICDS-774)

##### SUMMARY
This adds ability to search by app profiles which are present on the current app state instead of build profiles which is only available when a version is selected for an app during search.
[tentative to-do]
give user the ability to search for profiles that are now deleted

As a prep, 
Break down forms for managing releases by app profile from a single form for searching and adding to two forms instead on the same page itself and keep only necessary fields for each within them.

##### FEATURE FLAG
`RELEASE_BUILDS_PER_PROFILE`

##### PRODUCT DESCRIPTION
User would now see two forms instead of now. A field to search by app profile is added in the search form which replaces search by build profile for search, 
the process to add a new release restriction remains the same.

Before
<img width="1426" alt="Screenshot 2019-09-16 at 5 45 35 PM" src="https://user-images.githubusercontent.com/3864163/64957138-dec47380-d8a9-11e9-9cdf-5f9b842b2a58.png">

After
<img width="1414" alt="Screenshot 2019-09-16 at 5 48 08 PM" src="https://user-images.githubusercontent.com/3864163/64957283-36fb7580-d8aa-11e9-93be-3565818e21a9.png">


<img width="1419" alt="Screenshot 2019-09-16 at 5 44 30 PM" src="https://user-images.githubusercontent.com/3864163/64957151-e71cae80-d8a9-11e9-91ce-87297f66869a.png">

mkangia committed:
app build profiles to view

mkangia committed:
rename

mkangia committed:
break down forms

mkangia committed:
populate app profiles in HTML

mkangia committed:
fix search

Update DOM references
build_profile_id -> app_profile_id

mkangia committed:
populate values on search for app profile instead

mkangia committed:
use a blank option instead of placeholder for clarity

select2 provides a small "x" to clear which isn't very clear or good UX

mkangia committed:
Merge branch 'master' into mk/app-profile-breakdown-forms

mkangia committed:
stickler

mkangia committed:
combine iteration and rename

mkangia committed:
Merge branch 'mk/app-profile-breakdown-forms' of github.com:dimagi/commcare-hq into mk/app-profile-breakdown-forms

orangejenny commented:
Could you post a before & after screenshot?

mkangia committed:
remove redundant help text [skip ci]

mkangia commented:
@orangejenny added.

orangejenny reviewed (changes requested):
Made a couple of minor comments, but stopped reviewing because I don't understand the naming here. The description says 

> This adds ability to search by app profiles which are present on the current app state instead of build profiles which is only available when a version is selected for an app during search.

[This code](https://github.com/dimagi/commcare-hq/pull/25380/files#diff-d190c4633f5bc404aefdf995bee0cd04R136) sets `app_profiles` equal to build profiles. So when you say "app profiles" do you mean the "whichever build profiles are available on the current app version"?

mkangia commented:
> whichever build profiles are available on the current app version

@orangejenny that is correct.

mkangia commented on corehq/apps/domain/templates/domain/manage_releases_by_app_profile.html:
```diff
@@ -12,52 +12,63 @@
-            </table>
-          {% endif %}
-        </div>
+    <ul class="nav nav-tabs sticky-tabs" style="margin-bottom: 10px;">
```
my guess is that you added a restriction sometime before and now you have deleted that version from the app? May be do an iteration on `LatestEnabledBuildProfiles.objects.all()` and check when `obj.build is None`?

orangejenny replied:
Thank you, got it working now, there was indeed a bad restriction 🙂 

I looked at this locally and think putting `<div class="spacer"></div>` below the ul is the best thing too do - I had thought there were other tab-related styles that would take care of that padding, but it doesn't seem like it.

While looking at this locally, I realized you can remove the wrapping `container` and `row` divs, and also it'd be nice to remove the legends, since now you have the tabs providing the same title & guidance.

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/25380/commits/0921ff721c26a883f393ae468ea673986eb90d3f

mkangia committed:
ko-select2 -> hqwebapp-select2

mkangia committed:
add primary btn for forms [skip ci]

orangejenny commented:
@mkangia Thank you for explaining. What would you think of renaming `app_profile` and related identifiers either back to `build_profile` or to something like `current_app_build_profiles`? It's wordy, but I'm finding this PR quite difficult to read, especially since "app profile" already has a different meaning (profile.ccpr).

mkangia committed:
update html [skip ci]

- remove inline styling
- remove legends
- remove redundant container and row classes

mkangia committed:
rename

app profile -> app build profile
apps profiles -> build profiles per app

mkangia committed:
rename form prop

mkangia committed:
remove unused context

mkangia committed:
search element on parent form instead of direct search [skip ci]

mkangia commented:
@orangejenny apologies for not getting back on this for a while. Renamed and made some minor improvements

mkangia committed:
stickler [skip ci]

orangejenny commented on corehq/apps/domain/views/releases.py:
```diff
@@ -109,7 +109,7 @@ class ManageReleasesByAppProfile(BaseProjectSettingsView):
 
     @cached_property
-    def form(self):
+    def create_form(self):
```
Nitpick: how about `creation_form` so it doesn't sound like it does the creation? (I know it's a property, but still.)

orangejenny replied:
Hrm, I see other places in HQ use `create_form` in this way, though I still think it's confusing.

mkangia replied:
i think `create_form` is right.
Do you think `search_form` should be renamed as well?

esoergel replied:
FWIW I agree with Jenny on this one - I'd expect `create_form` to be a callable that returns a form, and `creation_form` to be a "noun" that can be referenced multiple times.

`search_form` sounds more like a noun than a verb to me though, so that name seems fine.

mkangia replied:
makes sense, [updated](https://github.com/dimagi/commcare-hq/pull/25380/commits/24d999800d32306b7f21a2e589633d8c2819c1d4)

mkangia commented:
@orangejenny is this gtg?

orangejenny commented:
@mkangia I haven't read through it thoroughly yet, should be able to do so today.

mkangia committed:
create -> creation

mkangia commented:
Sure, thanks @orangejenny . Looking forward to it.

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/widgets.js:
```diff
@@ -56,24 +56,23 @@ hqDefine("app_manager/js/widgets", [
-                    data: buildProfiles,
-                });
+                if (buildProfiles.length) {
+                    $('#build-profile-id-input').append(new Option(gettext('Select Profile'), '', true, true));
```
Nitpick: it's be more semantic to pass a `placeholder` option to the select2 call.

mkangia replied:
i remember we talked about that option on a PR earlier and then decided to keep this considering few factors. I am happy to link that if needed. So keeping this.

orangejenny replied:
Now that you mention it, that does sound familiar... :)

orangejenny commented on corehq/apps/domain/views/releases.py:
```diff
@@ -108,58 +109,63 @@ class ManageReleasesByAppProfile(BaseProjectSettingsView):
-                    'text': details['name'],
-                    'selected': build_profile_id == _id
-                } for _id, details in build_doc['build_profiles'].items()]
+        if build_profiles_per_app and app_id and app_id in build_profiles_per_app:
```
I think you can leave off the `and app_id`, assuming there wouldn't be a `None` in `build_profiles_per_app`.

mkangia replied:
it's coming from search and is a required field in the form, so ideally it should always be present but good to still check it.

mkangia committed:
Merge branch 'master' into mk/app-profile-breakdown-forms

mkangia commented:
ran preindex on production and india.
both took 5-10 mins
